context/context.go

```diff
@@ -2,56 +2,50 @@ package context
 
 import (
 	"context"
-	"errors"
 
 	"github.com/sirupsen/logrus"
 )
 
 type indexContext int
 
 const (
-	traceIDKey = iota
+	loggerKey = iota
 )
 
 var (
-	fieldLogger logrus.FieldLogger
+	baseLogger logrus.FieldLogger
 )
 
 func init() {
-	fieldLogger = logrus.New().WithFields(nil)
+	baseLogger = logrus.New().WithFields(nil)
 }
 
-func SetLogger(logger logrus.FieldLogger) {
-	fieldLogger = logger
+func SetBaseLogger(logger logrus.FieldLogger) {
+	baseLogger = logger
 }
 
 func WithTraceID(ctx context.Context, traceID string) context.Context {
-	if traceID != "" {
-		ctx = context.WithValue(ctx, traceIDKey, traceID)
+	if traceID == "" {
+		return ctx
 	}
+	ctx = logWithField(ctx, "traceID", traceID)
 	return ctx
 }
 
-func Logger(ctx context.Context) logrus.FieldLogger {
-	logger := fieldLogger
-
-	if id, err := traceID(ctx); err == nil {
-		logger = logger.WithField("traceID", id)
-	}
-
-	return logger
+func logWithField(ctx context.Context, key string, value interface{}) context.Context {
+	log := Logger(ctx).WithField(key, value)
+	return context.WithValue(ctx, loggerKey, log)
 }
 
-func traceID(ctx context.Context) (string, error) {
-	traceID, ok := ctx.Value(traceIDKey).(string)
+func Logger(ctx context.Context) logrus.FieldLogger {
+	log, ok := ctx.Value(loggerKey).(*logrus.Entry)
```

Comment:
We might not need to do that, since the library seems to handle a logger pool internally:
- https://github.com/sirupsen/logrus/blob/master/logger.go#L93-L97
- https://github.com/sirupsen/logrus/blob/master/logger.go#L78-L83